docs/architecture/adr-062-p2p-architecture.md

```diff
@@ -0,0 +1,531 @@
+
+## Alternative Approaches
+
+Several variations of the proposed design were considered, including e.g. calling interface methods instead of passing messages (like the current architecture), merging channels with streams, exposing the internal peer data structure to reactors, being message format-agnostic via arbitrary codecs, and so on. This design was chosen because it has very loose coupling, is simpler to reason about and more convenient to use, avoids race conditions and lock contention for internal data structures, gives reactors better control of message ordering and processing semantics, and allows for QoS scheduling and backpressure in a very natural way.
+
+There were also proposals to use LibP2P instead of maintaining our own P2P stack, which was rejected (for now) in [ADR 061](adr-061-p2p-refactor-scope.md).
+
+## Decision
+
+The P2P stack will be redesigned as a message-oriented architecture, primarily relying on Go channels for communication and scheduling. It will use IO stream transports to exchange raw bytes with individual peers, bidirectional peer-addressable channels to send and receive Protobuf messages, and a router to route messages between reactors and peers. Message passing is asynchronous with at-most-once delivery.
+
+## Detailed Design
+
+This ADR is primarily concerned with the architecture and interfaces of the P2P stack, not implementation details. Separate ADRs may be submitted for individual components, since implementation may be non-trivial. The interfaces described here should therefore be considered a rough architecture outline, not a complete and final design.
+
+Primary design objectives have been:
+
+* Loose coupling between components, for a simpler, more robust, and test-friendly architecture.
+* Pluggable transports (not necessarily networked).
+* Better scheduling of messages, with improved prioritization, backpressure, and performance.
+* Centralized peer lifecycle and connection management.
+* Better peer address detection, advertisement, and exchange.
+* Backwards compatibility at the wire-level with current P2P network protocols.
+
+The main abstractions in the new stack are:
+
+* `peer`: A node in the network, uniquely identified by a `PeerID` and stored in a `peerStore`.
+* `Transport`: An arbitrary mechanism to exchange bytes with a peer using IO `Stream`s across a `Connection`.
+* `Channel`: A bidirectional channel to asynchronously exchange Protobuf messages with peers addressed with `PeerID`.
+* `Router`: Maintains transport connections to relevant peers and routes channel messages.
+* Reactor: A design pattern loosely defined as "something which listens on a channel and reacts to messages".
+
+These abstractions and related concepts are illustrated in the following diagram and described in detail below.
+
+![P2P Architecture Diagram](img/adr-062-architecture.svg)
```

Comment:
Just to be clear: Is everything in this diagram that's _not_ "Peer B," "Peer C," or "Peer D" part of the same peer (presumably the unnamed "Peer A")?

Reply:
Yes - I guess that should be made explicit.